students/DrewSmith/lessons/lesson07/assignment/parallel.py

```diff
@@ -0,0 +1,277 @@
+        customer_id, first_name, last_name, home_address,phone_number, email_address,
+        status, credit_limit
+    '''
+    records_added = 0
+    batch_size = 120
+    customer_list = list()
+    database = cc.get_database()
+    pool = ThreadPool(processes=3)
+    for index, customer in enumerate(customers, start=1):
+        try:
+            customer['credit_limit'] = float(customer['credit_limit'])
+        except ValueError as err:
+            logging.error(err)
+            raise ValueError(f"Invalid value: record {index}: " \
+                f"invalid credit_limit: '{customer['credit_limit']}'")
+
+        if not isinstance(customer['status'], bool):
+            text = f"Invalid value: record {index}: status is not a bool: '{customer['status']}'"
+            logging.error(text)
+            raise ValueError(text)
+
+        customer_list.append(customer)
+        if index % batch_size == 0:
+            pool.apply_async(_insert_customer_list, (list(customer_list), database))
+            customer_list = list()
+            records_added += batch_size
+
+    if len(customer_list) > 0:
+        pool.apply_async(_insert_customer_list, (list(customer_list), database))
+        records_added += len(customer_list)
+    
+    pool.close()
+    pool.join()
+    
+    return records_added
+
+def _insert_customer_list(customer_list, database):
+    '''
+    Internal function to add customer list with error checking
+
+    :param customer_list: sequence of customer dictionaries to insert
+    :param database: database object for the atomic transaction
+    '''
+    with database.atomic():
+        try:
+            Customer.insert_many(customer_list).execute()
+        except pw.IntegrityError as error:
+            logging.error(f"Customer already exists: {str(error)}")
+            raise
+        else:
+            LOGGER.info(f"Inserted {len(customer_list)} customer(s)")
+
+def get_customers_record_count():
+    '''
+    Get the Customers count
+    '''
+    return Customer.select().count()
+
+
+def _database_setup():
+    '''
+    reset database
+    '''
+    cc.get_database().drop_tables([Customer])
+    cc.get_database().create_tables([Customer])
+
+    mongo = MongoDBConnection()
+    with mongo:
+        db = mongo.connection.get_database(name=config.DATABASE_NAME)
+        db.drop_collection("customers")
+        db.drop_collection("products")
+        db.drop_collection("rentals")
+
+def main():
+    _database_setup()
+    pool = ThreadPool(processes=2)
+
+    start = datetime.datetime.now()
+    parent_path = Path(__file__).parent
+
+    products_starting = get_prodcuts_total_record_count()
+    customers_starting = get_customers_record_count()
+
+    products_thread = pool.apply_async(import_product_data, (parent_path, "products.csv", "contacts.csv", "rentals.csv"))
+    customers_thread = pool.apply_async(bulk_add_customers, (convert_csv("customers.csv"), ))
+    pool.close()
+    pool.join()
```

Comment:
Interesting, so you are threading for mongo and sqlite, and then threading again for each of those threads? I'm not sure how this would affect things. I'd be curious how the sqlite and mongo do individually. Or for that matter, threading only the top level. Did you try any of those? I'm not sure how threading a thread goes, in general, but I'm definitely intrigued.